gateway/join.go

```diff
@@ -0,0 +1,254 @@
+	rx2Frequenecy    = 923300000 // Frequuency for rx2 window
+	rx2SF            = 12        // spreading factor for rx2 window
+	rx2Bandwidth     = 0x06      // 500k bandwidth
+)
+
+var errNoDevice = errors.New("received join request from unknown device")
+
+// network id for the device to identify the network.
+var netID = []byte{1, 2, 3}
+
+func (g *Gateway) handleJoin(ctx context.Context, payload []byte) error {
+	jr, device, err := parseJoinRequestPacket(payload, g.devices)
+	if err != nil {
+		return err
+	}
+
+	joinAccept, err := device.GenerateJoinAccept(ctx, jr)
+	if err != nil {
+		return err
+	}
+
+	txPkt := C.struct_lgw_pkt_tx_s{
+		freq_hz:    C.uint32_t(rx2Frequenecy),
+		tx_mode:    C.uint8_t(0), // immediate mode
+		rf_chain:   C.uint8_t(0),
+		rf_power:   C.int8_t(26),    // tx power in dbm
+		modulation: C.uint8_t(0x10), // LORA modulation
+		bandwidth:  C.uint8_t(rx2Bandwidth),
+		datarate:   C.uint32_t(rx2SF),
+		coderate:   C.uint8_t(0x01), // code rate 4/5
+		invert_pol: C.bool(true),    // Downlinks are always reverse polarity.
+		size:       C.uint16_t(len(joinAccept)),
+	}
+
+	var cPayload [256]C.uchar
+	for i, b := range joinAccept {
+		cPayload[i] = C.uchar(b)
+	}
+	txPkt.payload = cPayload
+
+	// send on rx2 window - opens 6 seconds after join request.
+	if !utils.SelectContextOrWait(ctx, time.Second*joinRx2WindowSec) {
+		return nil
+	}
+
+	// lock so there is not two sends at the same time.
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	errCode := int(C.send(&txPkt))
+	if errCode != 0 {
+		return errors.New("failed to send join accept packet")
+	}
+
+	return nil
+}
+
+// payload of join request consists of
+// | MHDR | JOIN EUI | DEV EUI  |   DEV NONCE  | MIC   |
+// | 1 B  |   8 B    |    8 B   |     2 B      |  4 B  |
+func parseJoinRequestPacket(payload []byte, devices map[string]*Device) (JoinRequest, *Device, error) {
+	var joinRequest JoinRequest
+
+	// everything in the join request payload is little endian
+	joinRequest.joinEUI = payload[1:9]
+	joinRequest.devEUI = payload[9:17]
+	joinRequest.devNonce = payload[17:19]
+	joinRequest.mic = payload[19:23]
+
+	matched := &Device{}
+
+	// device.devEUI is in big endian - reverse to compare and find device.
+	devEUIBE := reverseByteArray(joinRequest.devEUI)
+
+	// match the dev eui to gateway device
+	for _, device := range devices {
+		if bytes.Equal(device.devEui, devEUIBE) {
+			matched = device
+		}
+	}
+
+	if matched.name == "" {
+		return JoinRequest{}, nil, errNoDevice
+	}
+
+	err := validateMIC(types.AES128Key(matched.AppKey), payload)
+	if err != nil {
+		return JoinRequest{}, nil, err
+	}
+
+	return joinRequest, matched, nil
+
+}
+
+// Format of Join Accept message:
+// | MHDR | JOIN NONCE | NETID |   DEV ADDR  | DL | RX DELAY |   CFLIST   | MIC  |
+// | 1 B  |     3 B    |   3 B |     4 B     | 1B |    1B    |  0 or 16   | 4 B  |
+func (d *Device) GenerateJoinAccept(ctx context.Context, jr JoinRequest) ([]byte, error) {
```

Comment:
Any reason this is the only exported function in this diff?

Reply:
no ill unexport